homeassistant/components/yamaha_musiccast/media_player.py

```diff
@@ -1,291 +1,371 @@
+        if self.coordinator.data.zones[self._zone_id].power == "on":
+            if self._is_netusb and self.coordinator.data.netusb_playback == "pause":
+                return STATE_PAUSED
+            if self._is_netusb and self.coordinator.data.netusb_playback == "stop":
+                return STATE_IDLE
+            return STATE_PLAYING
+        return STATE_OFF
 
     @property
-    def is_volume_muted(self):
-        """Boolean if volume is currently muted."""
-        return self.mute
+    def volume_level(self):
+        """Return the volume level of the media player (0..1)."""
+        volume = self.coordinator.data.zones[self._zone_id].current_volume
+        return (volume - self._volume_min) / (self._volume_max - self._volume_min)
 
     @property
-    def volume_level(self):
-        """Volume level of the media player (0..1)."""
-        return self.volume
+    def is_volume_muted(self):
+        """Return boolean if volume is currently muted."""
+        return self.coordinator.data.zones[self._zone_id].mute
 
     @property
-    def supported_features(self):
-        """Flag of features that are supported."""
-        return SUPPORTED_FEATURES
+    def shuffle(self):
+        """Boolean if shuffling is enabled."""
+        return (
+            self.coordinator.data.netusb_shuffle == "on" if self._is_netusb else False
+        )
 
     @property
-    def source(self):
-        """Return the current input source."""
-        return self._source
+    def sound_mode(self):
+        """Return the current sound mode."""
+        return self.coordinator.data.zones[self._zone_id].sound_program
 
     @property
-    def source_list(self):
-        """List of available input sources."""
-        return self._source_list
+    def sound_mode_list(self):
+        """Return a list of available sound modes."""
+        return self.coordinator.data.zones[self._zone_id].sound_program_list
 
-    @source_list.setter
-    def source_list(self, value):
-        """Set source_list attribute."""
-        self._source_list = value
+    @property
+    def device_class(self):
+        """Return the device class of the media player."""
+        return self._device_class
 
     @property
-    def media_content_type(self):
-        """Return the media content type."""
-        return MEDIA_TYPE_MUSIC
+    def zone(self):
+        """Return the zone of the media player."""
+        return self._zone_id
 
     @property
-    def media_duration(self):
-        """Duration of current playing media in seconds."""
-        return self.media_status.media_duration if self.media_status else None
+    def unique_id(self) -> str:
+        """Return the unique ID for this media_player."""
+        macs = self.coordinator.data.mac_addresses
+        return f"{macs}_{self._zone_id}"
+
+    async def async_turn_on(self):
+        """Turn the media player on."""
+        await self.coordinator.musiccast.turn_on(self._zone_id)
+        self.schedule_update_ha_state()
```

Comment:
Make sure to use the async api in async context.
```suggestion
        self.async_write_ha_state()
```